ykcompile/src/lib.rs

```diff
@@ -73,7 +78,14 @@ impl CompiledTrace {
         // For now a compiled trace always returns whatever has been left in register RAX. We also
         // assume for now that this will be a `u64`.
         let func: fn() -> u64 = unsafe { mem::transmute(self.mc.ptr(dynasmrt::AssemblyOffset(0))) };
-        func()
+        self.exec_trace(func)
+    }
+
+    /// Jump to JITted code. This is a separate unmangled function to make it easy to set a
+    /// debugger breakpoint right before entering the trace.
+    #[no_mangle]
+    fn exec_trace(&self, t_fn: fn() -> u64) -> u64 {
```

Comment:
1. You can use `rbreak exec_trace` (gdb) or `break set -r exec_trace` (lldb) to set a breakpoint on any symbol containing `exec_trace`.
2. LLVM can inline this function and I expect it to do so in release mode.

Reply:
> You can use rbreak exec_trace (gdb) or break set -r exec_trace (lldb) to set a breakpoint on any symbol containing exec_trace.

Didn't know that! Thanks!

> LLVM can inline this function and I expect it to do so in release mode.

Yeah. I realised that and initially I had a `#[inline(never]` on this function, but realised that it wasn't necessary for us right now: we explicitly turn off the optimiser for now because we don't want LLVM to re-order blocks.

Reply:
Would it be possible to enable optimizations for yorick itself, but keep it disabled for user code? That would speed up the tracing and jitting while keeping LLVM from reordering blocks in the user code.

Reply:
It might be, yeah. Might be able to filter based on crate, but we may have to move the tests to another crate so that they don't get optimised (they emulate user code).

Reply:
I've made an issue for this:
https://github.com/softdevteam/ykrustc/issues/105

Reply:
It seems that optimizations are disabled by default for tests: https://doc.rust-lang.org/cargo/reference/profiles.html#test